util/test/demos/gl/gl_pixelhistory_formats.cpp

```diff
@@ -0,0 +1,245 @@
+
+    glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, sizeof(DefaultA2V), (void *)(0));
+    glVertexAttribPointer(1, 4, GL_FLOAT, GL_FALSE, sizeof(DefaultA2V), (void *)(sizeof(Vec3f)));
+    glVertexAttribPointer(2, 2, GL_FLOAT, GL_FALSE, sizeof(DefaultA2V),
+                          (void *)(sizeof(Vec3f) + sizeof(Vec4f)));
+
+    glEnableVertexAttribArray(0);
+    glEnableVertexAttribArray(1);
+    glEnableVertexAttribArray(2);
+
+    GLuint program = MakeProgram(common + vertex, common + pixel);
+
+    GLenum colorFormats[] = {
+        GL_RGBA8,      GL_RGBA16,         GL_RGBA16F, GL_RGBA32F,  GL_RGBA8I,  GL_RGBA8UI,
+        GL_RGBA16I,    GL_RGBA16UI,       GL_RGBA32I, GL_RGBA32UI, GL_RGB8,    GL_RGB16,
+        GL_RGB16F,     GL_RGB32F,         GL_RGB8I,   GL_RGB8UI,   GL_RG8,     GL_RG16,
+        GL_RG16F,      GL_RG32F,          GL_RG8I,    GL_RG8UI,    GL_R8,      GL_R16,
+        GL_R16F,       GL_R32F,           GL_R8I,     GL_R8UI,     GL_RGB5_A1, GL_RGB10_A2,
+        GL_RGB10_A2UI, GL_R11F_G11F_B10F, GL_RGB565};
+
+    GLenum depthFormats[] = {GL_DEPTH_COMPONENT16,  GL_DEPTH_COMPONENT24, GL_DEPTH_COMPONENT32,
+                             GL_DEPTH_COMPONENT32F, GL_DEPTH24_STENCIL8,  GL_DEPTH32F_STENCIL8};
+
+    constexpr size_t colorFormatSize = sizeof(colorFormats) / sizeof(GLenum);
+    constexpr size_t depthFormatSize = sizeof(depthFormats) / sizeof(GLenum);
+
+    GLuint colorTextures[colorFormatSize];
+    GLuint multisampledColorTextures[colorFormatSize];
+    GLuint depthTextures[depthFormatSize];
+
+    for(size_t i = 0; i < colorFormatSize; ++i)
+    {
+      colorTextures[i] = MakeTexture();
+      multisampledColorTextures[i] = MakeTexture();
+      glBindTexture(GL_TEXTURE_2D, colorTextures[i]);
+      glTexStorage2D(GL_TEXTURE_2D, 1, colorFormats[i], screenWidth, screenHeight);
+      glBindTexture(GL_TEXTURE_2D_MULTISAMPLE, multisampledColorTextures[i]);
+      glTexStorage2DMultisample(GL_TEXTURE_2D_MULTISAMPLE, 2, colorFormats[i], screenWidth,
+                                screenHeight, GL_TRUE);
+    }
+
+    for(size_t i = 0; i < depthFormatSize; ++i)
+    {
+      depthTextures[i] = MakeTexture();
+      glBindTexture(GL_TEXTURE_2D, depthTextures[i]);
+      glTexStorage2D(GL_TEXTURE_2D, 1, depthFormats[i], screenWidth, screenHeight);
+    }
+
+    GLuint fbo = MakeFBO();
+    glBindFramebuffer(GL_FRAMEBUFFER, fbo);
+
+    glDepthFunc(GL_ALWAYS);
+    glEnable(GL_DEPTH_TEST);
+    glDepthMask(GL_TRUE);
+
+    glStencilFunc(GL_ALWAYS, 0xcc, 0xff);
+    glStencilOp(GL_REPLACE, GL_REPLACE, GL_REPLACE);
+    glEnable(GL_STENCIL_TEST);
+    glStencilMask(0xff);
+
+    while(Running())
+    {
+      glBindFramebuffer(GL_FRAMEBUFFER, fbo);
+
+      for(size_t i = 0; i < colorFormatSize; ++i)
+      {
+        glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D,
+                               colorTextures[i], 0);
+
+        if(i >= depthFormatSize)
```

Comment:
Since the number of combinations isn't too high I think it would be nice to do an outer loop over depth formats (including none) and an inner loop over depth formats. In fact I think it would be reasonably to have an even further outer loop over multisampled vs not. That still totals to under 500 combinations which is entirely reasonable.

The chances of a bug being specific to one particular combination are low, but the cost of testing exhaustively is pretty low.